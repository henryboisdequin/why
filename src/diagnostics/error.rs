use std::backtrace::Backtrace;

/// Error type which holds all of the errors information.
#[derive(PartialEq, Debug)]
pub struct Error<'err> {
    /// The main message/cause of the error.
    pub message: &'err str,
    /// Any help messages which could point the user to the right direction.
    pub help_msgs: Option<Vec<&'err str>>,
    /// Any notes which could give the user additional information.
    pub note_msgs: Option<Vec<&'err str>>,
    /// The error code of the current error.
    pub code: Option<usize>,
    /// The file in which the error occurred.
    pub file: Option<&'err str>,
    /// The line in which the error occurred.
    pub line: Option<usize>,
    /// Whether to display a stack backtrace or not.
    pub debug: bool,
}

impl<'err> Error<'err> {
    /// Returns a new default `Error`.
    pub fn new(debug: bool) -> Self {
        Self {
            message: "",
            help_msgs: None,
            note_msgs: None,
            code: None,
            file: None,
            line: None,
            debug,
        }
    }

    /// Adds a main label to the error.
    pub fn span_label(&mut self, code: Option<usize>, message: &'err str) {
        if let Some(err_code) = code {
            self.code = Some(err_code);
        }

        self.message = message;
    }

    /// Adds a note to the error.
    pub fn note(&mut self, note: &'err str) {
        if let Some(notes_msgs) = self.note_msgs.as_mut() {
            notes_msgs.push(note);
            self.note_msgs = Some(notes_msgs.to_vec());
        } else {
            self.note_msgs = Some(vec![note]);
        }
    }

    /// Adds a help message to the error.
    pub fn help(&mut self, help_msg: &'err str) {
        if let Some(help_msgs) = self.help_msgs.as_mut() {
            help_msgs.push(help_msg);
            self.help_msgs = Some(help_msgs.to_vec());
        } else {
            self.help_msgs = Some(vec![help_msg]);
        }
    }

    /// Adds the file and line in which this error occurred.
    pub fn add_file_and_line(&mut self, file: &'err str, line: usize) {
        self.file = Some(file);
        self.line = Some(line);
    }

    /// Shows the backtrace of the error.
    pub fn show_backtrace(self) -> Option<Backtrace> {
        if self.debug {
            Some(Backtrace::capture())
        } else {
            None
        }
    }
}
