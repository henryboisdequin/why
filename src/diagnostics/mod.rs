/// Responsible for the diagnostic builder.
pub mod builder;
/// Holds and manages the errors in `why`.
pub mod error;
/// Implements display for all of the structs in `why`.
pub mod fmt;
