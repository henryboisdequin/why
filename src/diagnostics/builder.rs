use super::error::Error;

/// Stores all errors the diagnostics in your program.
#[derive(PartialEq, Debug)]
pub struct DiagnosticBuilder<'err> {
    /// Vector of all the errors in your program.
    pub errors: Vec<Error<'err>>,
    /// How many errors have occurred.
    pub error_count: usize,
}

impl<'err> DiagnosticBuilder<'err> {
    /// Returns a new `DiagnosticBuilder` with an empty vector of errors.
    pub fn new() -> Self {
        Self {
            errors: Vec::new(),
            error_count: 0,
        }
    }

    /// Adds an error to the error vector.
    pub fn add_err(&mut self, error: Error<'err>) {
        self.errors.push(error);
        self.error_count += 1;
    }

    /// Emits all the errors by calling `print_all_errs`.
    pub fn emit(self) {
        self.print_all_errs();
    }

    /// Prints all of the errors by running `display`.
    fn print_all_errs(self) {
        for error in &self.errors {
            println!("{}", error);
            if self.error_count == 1 {
                println!("aborting due to the previous error")
            } else {
                println!("aborting due to {} previous errors", self.error_count);
            }
        }
    }
}
