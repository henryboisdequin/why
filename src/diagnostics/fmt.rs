use super::error::Error;
use std::fmt::{Display, Formatter, Result as FmtResult};

impl Display for Error<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        let mut diagnostic = String::new();

        // The top part of the error.
        // E.g. (with code)
        // ```
        // error[101]: Mismatched types
        // ```
        // E.g. (without code)
        // ```
        // error: Mismatched types
        // ```
        if let Some(code) = self.code {
            diagnostic.push_str(&format!("error[{}]: {}\n", code, self.message));
        } else {
            diagnostic.push_str(&format!("error: {}\n", self.message));
        }

        // Adds file and line information to the diagnostic
        if let (Some(file), Some(line)) = (self.file, self.line) {
            diagnostic.push_str(&format!("cause: {}:{}\n", file, line));
        }

        // The helpful notes of the given error.
        // E.g.
        // ```
        // note: `Foo` does not implement `Clone`
        // ```
        if let Some(note_msgs) = self.note_msgs.clone() {
            for note_msg in note_msgs {
                diagnostic.push_str(&format!("note: {}\n", note_msg));
            }
        }

        // The help messages of the given error.
        // E.g.
        // ```
        // help: implement `Clone` for `Foo`. add `#[derive(Clone)]` above the definition of `Foo`
        // ```
        if let Some(help_msgs) = self.help_msgs.clone() {
            for help_msg in help_msgs {
                diagnostic.push_str(&format!("help: {}\n", help_msg));
            }
        }

        // Write the whole diagnostic to the formatter
        write!(f, "{}", diagnostic)?;
        Ok(())
    }
}
