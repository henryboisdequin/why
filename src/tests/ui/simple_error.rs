use why::diagnostics::{builder::DiagnosticBuilder, error::Error};

fn main() {
    let mut builder = DiagnosticBuilder::new();
    let mut error = Error::new(false);
    error.span_label(Some(101), "fail");
    error.note("it failed");
    error.add_file_and_line(file!(), line!() as usize);
    builder.add_err(error);
    builder.emit();
}
