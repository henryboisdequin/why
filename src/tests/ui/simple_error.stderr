error[101]: fail
cause: /Users/henryboisdequin/Desktop/why/src/tests/ui/simple_error.rs:8
note: it failed

aborting due to the previous error
